### PR TITLE
Highlighted branch names in log and status views

### PR DESCRIPTION
**Pull request by GriffinSchneider**
image

This PR adds branch/ref names to the status view, and highlights ref names in the status and log views using semantic highlighting based on the `View` tree.

The majority of the work here was in refactoring the `View` system. Previously, every `View` implicitly rendered itself on its own line, but since we want to highlight specific parts of some lines, this had to change. To make the change easier, subclasses of `View` still put themselves on a new line, but subclasses of `TextView` don't do that anymore. There were various views that extended `TextView` (`HunkView`, for example), and those have been changed to extending `View` and adding `TextView`s as subviews so they retain their previous newline behavior.

As far as the actual highlighting, there's a new view `TokenView` to indicate that some text should be highlighted as a semantic token. Then we register a `DocumentSemanticTokensProvider` that crawls the `View` tree, finds any `TokenView`s, and translates them into semantic tokens for highlighting.

There's also some related work in here to add ref names to commits in the status view like Emacs magit - basically I just look through the `magitState` for any refs matching commits we're displaying and render them as `TokenView`s.

I tried to follow the general code style of the projet, but I'm happy to make any changes if you've got comments.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -380,6 +380,26 @@
...
+        "id": "magit-remote-ref-name",
+        "superType": "label",
+        "description": "Name of a git ref from a remote."
+      }
+    ],
+    "semanticTokenScopes": [
+      {
+        "scopes": {
+          "magit-ref-name": ["entity.name.type.template"],
+          "magit-remote-ref-name": ["variable.other.constant"]
```
Comment (GriffinSchneider) [inline package.json:399]: I don't really know what I'm doing with these, I just ended up picking some values that happen to look OK with my theme ¯\\\_(ツ)\_/¯

**Comment by kahole** [comment]
Thanks for submitting a PR on this, it looks good so far.
I did encounter an issue, however. This implementation seems to break both folding and highlighting of sections when the cursor is at character number zero. This will need to be adressed.

To reproduce:

Press <tab> when on character zero for a section header. (e.g. Unstaged changes (1))
The section above gets folded.

**Comment by GriffinSchneider** [comment]
Thanks for pointing that out, I think I've fixed the problem. I had an off-by-one error in how `View`s calculated their `range` when the `View`'s content ended in a newline, which caused the first character of the next `View` to count as part of the previous `View`. I tested a bunch of different combinations of folding things and cursor positions and e22ee2ebe89615bb50e2c66a01885ebc5d568042 seems to get everything back to working properly.

**Comment by kahole** [comment]
@GriffinSchneider 
After merging, I found some new issues that this PR introduces.

With this cursor position, the view shouldn't be highlighted.

Screenshot 2020-07-14 at 18 40 02

Also there is an extra new line at the bottom of the status view.

**Comment by GriffinSchneider** [comment]
Ah, I didn't realize that was a regression - Emacs magit actually does both of those things, but I do think the way it used to work was better. https://github.com/kahole/edamagit/pull/58 should resolve both those issues.

**Pull request merged**